lib/octopus_serverspec_extensions/type/octopus_deploy_smtp_config.rb

```diff
@@ -0,0 +1,109 @@
+require 'serverspec/type/base'
+require 'net/http'
+require 'json'
+
+module Serverspec::Type
+  class OctopusDeploySmtpConfig < Base
+    @serverUrl = nil
+    @apiKey = nil
+    @smtpConfig = nil
+
+    def initialize(*url_and_api_key)
+      serverUrl, apiKey = get_octopus_creds(url_and_api_key)
+
+      @name = "Octopus Deploy SMTP Config #{serverUrl}"
+      @runner = Specinfra::Runner
+      @serverUrl = serverUrl
+      @apiKey = apiKey
+
+      if serverUrl.nil?
+        serverUrl = get_env_var('OCTOPUS_CLI_SERVER').chomp('/') # removes trailing slash if present
+      end
+
+      if apiKey.nil?
+        apiKey = get_env_var('OCTOPUS_CLI_API_KEY')
+      end
+
+      # is it still nil?
+      if serverUrl.nil?
+        raise "'serverUrl' was not provided. Unable to connect to Octopus server to validate configuration."
+      end
+      if apiKey.nil?
+        raise "'apiKey' was not provided. Unable to connect to Octopus server to validate configuration."
+      end
+
+      @smtpConfig = get_smtp_config_via_api(serverUrl, apiKey)
+    end
+
+    def configured?
+      url = "#{@serverUrl}/api/smtpconfiguration/isconfigured?api-key=#{@apiKey}"
+      begin
+        resp = Net::HTTP.get_response(URI.parse(url))
+        body = JSON.parse(resp.body)
+        smtp = body unless body.nil?
+      rescue => e
+        raise "get_smtp_config_via_api: Unable to connect to #{url}: #{e}"
+      end
+
+      smtp["IsConfigured"]
+
+    end
+
+    def using_ssl?
+      false if @smtpConfig.nil?
+      @smtpConfig["EnableSsl"]
+    end
+
+    def on_port?(port_number)
+      false if @smtpConfig.nil?
+      @smtpConfig["SmtpPort"] == port_number
+    end
+
+    def on_host?(hostname)
+      false if @smtpConfig.nil?
+      @smtpConfig["SmtpHost"] == hostname
+    end
+
+    def using_credentials?(username)
+      # we can't test the password, but we can check the username
+      false if @smtpConfig.nil?
+      @smtpConfig["SmtpLogin"] == username && @smtpConfig["SmtpPassword"]["HasValue"]
+    end
+
+    def has_from_address?(from_address)
+      false if @smtpConfig.nil?
+      @smtpConfig["SendEmailFrom"] == from_address
+    end
+  end
+
+
+    def octopus_deploy_smtp_config(*url_and_api_key)
```

Comment:
indenting and that double `end` on line 77 is throwing me.
is that line 77 the closing of the class, and these 2 are the static constructors or something?
Maybe a comment to that effect?

Reply:
Yeah, they're methods on the module, that allow you to instantiate the subclass. Effectively static constructors.

That double end threw me badly at one point, couldn't figure out why my class wasn't loading for ages

Reply:
Indenting fixed